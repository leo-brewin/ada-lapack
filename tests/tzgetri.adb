with Ada.Text_IO; 
with Ada.Text_IO.Complex_IO;
with Ada.Numerics.Generic_Real_Arrays;
with Ada.Numerics.Generic_Complex_Types;
with Ada.Numerics.Generic_Complex_Arrays;
with Ada.Numerics.Generic_Elementary_Functions;
with Ada.Numerics.Generic_Complex_Elementary_Functions;
with Ada_Lapack;

use Ada.Text_IO;

procedure tzgetri is
   
   type Real is digits 18;

   package Real_Arrays     is new Ada.Numerics.Generic_Real_Arrays (Real);
   package Complex_Types   is new Ada.Numerics.Generic_Complex_Types (Real);
   package Complex_Arrays  is new Ada.Numerics.Generic_Complex_Arrays (Real_Arrays, Complex_Types);
   
   package Real_Maths      is new Ada.Numerics.Generic_Elementary_Functions (Real);
   package Complex_Maths   is new Ada.Numerics.Generic_Complex_Elementary_Functions (Complex_Types);

   package Real_IO         is new Ada.Text_IO.Float_IO (Real);
   package Integer_IO      is new Ada.Text_IO.Integer_IO (Integer);
   package Complex_IO      is new Ada.Text_IO.Complex_IO (Complex_Types);

   package Lapack          is new Ada_Lapack(Real, Complex_Types, Real_Arrays, Complex_Arrays);
      
   use Lapack;
   
   use Real_Arrays;
   use Complex_Types;
   use Complex_Arrays;
   
   use Real_IO;
   use Integer_IO;
   use Complex_IO;
   
   use Real_Maths;
   use Complex_Maths;
   
   matrix      : Complex_Matrix (1..4,1..4);
   matrix_copy : Complex_Matrix (1..4,1..4);
   
   matrix_rows : Integer := Matrix'Length (1);
   matrix_cols : Integer := Matrix'Length (2);
         
   pivots : Integer_Vector (1..matrix_rows);

   short_vector : Complex_Vector (1..1);
   
   return_code : Integer;

begin
   
   matrix :=( ((  1.23, -5.50), (  7.91, -5.38), ( -9.80, -4.86), ( -7.32,  7.57)),
              (( -2.14, -1.12), ( -9.92, -0.79), ( -9.18, -1.12), (  1.37,  0.43)),
              (( -4.30, -7.10), ( -6.47,  2.52), ( -6.51, -2.67), ( -5.86,  7.38)),
              ((  1.27,  7.29), (  8.90,  6.92), ( -8.82,  1.25), (  5.41,  5.37)) );
   
   matrix_copy := matrix;
   
   GETRF ( A       => matrix,
           M       => matrix_rows,
           N       => matrix_rows,
           LDA     => matrix_rows,
           IPIV    => pivots,
           INFO    => return_code );
        
   GETRI ( A       => matrix,
           N       => matrix_rows,
           LDA     => matrix_rows,
           IPIV    => pivots,
           WORK    => short_vector,
           LWORK   => -1,
           INFO    => return_code);

   declare
      work_vector_rows : Integer := Integer( short_vector(1).Re );
      work_vector : Complex_Vector (1 .. work_vector_rows);
   begin

      GETRI ( A       => matrix,
              N       => matrix_rows,
              LDA     => matrix_rows,
              IPIV    => pivots,
              WORK    => work_vector,
              LWORK   => work_vector_rows,
              INFO    => return_code);
               
   end;

   if (return_code /= 0) then

      Put_line ("The matrix is probably singular.");

   else
      
      Put_line ("Matrix");
      for i in matrix_copy'range(1) loop
         for j in matrix_copy'range(2) loop
            put(matrix_copy(i,j),3,3,0);
            put(" ");
         end loop;
         new_line;
      end loop;
      new_line;
         
      Put_line ("Inverse");
      for i in matrix'range(1) loop
         for j in matrix'range(2) loop
            put(matrix(i,j),3,3,0);
            put(" ");
         end loop;
         new_line;
      end loop;
         
   end if;
      
end tzgetri;
