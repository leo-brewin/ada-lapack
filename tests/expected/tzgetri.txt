Matrix
(  1.230, -5.500) (  7.910, -5.380) ( -9.800, -4.860) ( -7.320,  7.570) 
( -2.140, -1.120) ( -9.920, -0.790) ( -9.180, -1.120) (  1.370,  0.430) 
( -4.300, -7.100) ( -6.470,  2.520) ( -6.510, -2.670) ( -5.860,  7.380) 
(  1.270,  7.290) (  8.900,  6.920) ( -8.820,  1.250) (  5.410,  5.370) 

Inverse
( -0.113, -0.054) ( -0.086,  0.011) (  0.127,  0.189) (  0.155, -0.034) 
(  0.057,  0.057) (  0.006,  0.013) ( -0.027, -0.098) ( -0.041, -0.034) 
( -0.059, -0.028) ( -0.093,  0.025) (  0.042,  0.036) (  0.014,  0.012) 
( -0.102,  0.067) (  0.021,  0.126) (  0.126, -0.095) (  0.039, -0.114) 
